### PR TITLE
Update phpstan-codeigniter and fix errors on Modules

### PR DESCRIPTION
**Pull request by paulbalandan**
**Description**
Explain what you have changed, and why.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Events/Events.php [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 namespace CodeIgniter\Events;
 
-use Config\Modules;
+use CodeIgniter\Modules\Modules;
```
Comment (paulbalandan) [inline system/Events/Events.php:14]:
```suggestion
use Config\Modules;
```

**Comment by paulbalandan** [comment]
I don't understand the failure on `testLoad`. using `dd` shows that it is `http://test.example.jp/this-is-test/`

**Comment by kenjis** [comment]
@paulbalandan It is a bug. I sent a PR #8037

**Comment by MGatner** [comment]
(Copying my comment from the other PR)

My only qualm with this new rule is that it hides dependencies from Deptrac. I think Config is fine as a globally-available dependency so I'm not too worried about it, but I did like the easily-built dependency graph.

**Comment by kenjis** [comment]
Is there any benefit in changing to short names in this main repository?

Another drawback is that completion in the IDE will not work.
If we use `::class`, IDE can autocomplete.
Screenshot 2023-10-12 20 16 50

File: system/CodeIgniter.php [hunk 2]
```diff
@@ -31,11 +31,10 @@
 use CodeIgniter\Router\Router;
 use Config\App;
 use Config\Cache;
-use Config\Feature;
 use Config\Kint as KintConfig;
 use Config\Services;
 use Exception;
-use Kint;
+use Kint\Kint;
```
[the remark this answers is not in the document]
Comment (kenjis) [reply system/CodeIgniter.php:37]: I think `Kint\Kint` is correct. `Kint` is a class alias.

**Comment by kenjis** [comment]
@paulbalandan By the way, why is `::class` call discouraged?
The only case this becomes a problem is if devs want to change the class in a module to be loaded at will.

**Comment by paulbalandan** [comment]
My inspiration for this is when I remember you made a PR on shield changing from `config(AuthJWT::class)` to `config('AuthJWT')` to allow overriding. So I thought why not made this a PHPStan rule.

**Comment by kenjis** [comment]
Yes, this is good for libraries (modules). Because `config(\CodeIgniter\Shield\Config\Auth::class)` do not load `\Config\Auth` (now by default).

But if there is `\Config\Foo` and devs use it, `config(\Config\Foo::class)` is no problem.

**Comment by paulbalandan** [comment]
Ok. So `Config\` should be the "champion". In cases where it is module-specific, that needs to be flagged. Hmm, I think that's doable.

**Comment by kenjis** [comment]
```
There were 2 errors:

1) CodeIgniter\Config\BaseConfigTest::testNotEnabled
RuntimeException: Registrars must return an array of properties and their values.

/home/runner/work/CodeIgniter4/CodeIgniter4/system/Config/BaseConfig.php:233
/home/runner/work/CodeIgniter4/CodeIgniter4/system/Config/BaseConfig.php:91
/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Config/BaseConfigTest.php:273

2) CodeIgniter\Config\BaseConfigTest::testDidDiscovery
RuntimeException: Registrars must return an array of properties and their values.

/home/runner/work/CodeIgniter4/CodeIgniter4/system/Config/BaseConfig.php:233
/home/runner/work/CodeIgniter4/CodeIgniter4/system/Config/BaseConfig.php:91
/home/runner/work/CodeIgniter4/CodeIgniter4/tests/system/Config/BaseConfigTest.php:288

ERRORS!
Tests: 548, Assertions: 930, Errors: 2.
```
https://github.com/codeigniter4/CodeIgniter4/actions/runs/6596168617/job/17924550098?pr=8036

File: system/Config/BaseConfig.php [hunk 3]
```diff
@@ -82,7 +82,7 @@ public static function __set_state(array $array)
      */
     public function __construct()
     {
-        static::$moduleConfig = config(Modules::class);
+        static::$moduleConfig = new Modules();
```
Comment (kenjis) [inline system/Config/BaseConfig.php:85]: `BaseConfig` does not share the `Config\Modules`.
So there is no way to change the `Config\Modules` values during testing.

https://github.com/codeigniter4/CodeIgniter4/blob/4c09ca3cf8ab28d6df03222710a73da386dfa80b/tests/system/Config/BaseConfigTest.php#L268-L281

Comment (kenjis) [reply system/Config/BaseConfig.php:85]: The fix will be something like this, but it is a breaking change.
```diff
diff --git a/system/Config/BaseConfig.php b/system/Config/BaseConfig.php
index b4b9a4a213..2d0a0e9fe6 100644
--- a/system/Config/BaseConfig.php
+++ b/system/Config/BaseConfig.php
@@ -80,9 +80,9 @@ class BaseConfig
      *
      * The "shortPrefix" is the lowercase-only config class name.
      */
-    public function __construct()
+    public function __construct(?Modules $config = null)
     {
-        static::$moduleConfig = new Modules();
+        static::$moduleConfig = $config ?? new Modules();
 
         if (! static::$override) {
             return;
diff --git a/tests/system/Config/BaseConfigTest.php b/tests/system/Config/BaseConfigTest.php
index 2da390392e..4c66673189 100644
--- a/tests/system/Config/BaseConfigTest.php
+++ b/tests/system/Config/BaseConfigTest.php
@@ -12,6 +12,7 @@
 namespace CodeIgniter\Config;
 
 use CodeIgniter\Test\CIUnitTestCase;
+use Config\Modules;
 use Encryption;
 use RegistrarConfig;
 use RuntimeException;
@@ -267,10 +268,10 @@ final class BaseConfigTest extends CIUnitTestCase
 
     public function testNotEnabled(): void
     {
-        $modulesConfig          = config('Modules');
+        $modulesConfig          = new Modules();
         $modulesConfig->enabled = false;
 
-        $config              = new RegistrarConfig();
+        $config              = new RegistrarConfig($modulesConfig);
         $config::$registrars = [];
         $expected            = $config::$registrars;
 
@@ -282,10 +283,10 @@ final class BaseConfigTest extends CIUnitTestCase
 
     public function testDidDiscovery(): void
     {
-        $modulesConfig          = config('Modules');
+        $modulesConfig          = new Modules();
         $modulesConfig->enabled = true;
 
-        $config              = new RegistrarConfig();
+        $config              = new RegistrarConfig($modulesConfig);
         $config::$registrars = [];
         $this->setPrivateProperty($config, 'didDiscovery', false);
 
```

Comment (MGatner) [reply system/Config/BaseConfig.php]: We can't modify that constructor. What about a static `@testTag` setter?

Comment (paulbalandan) [reply system/Config/BaseConfig.php]: Okay. I'll change.

Comment (paulbalandan) [reply system/Config/BaseConfig.php]: Would it be okay to change `static::$moduleConfig` as nullable? phpstan complains on `??=`

Comment (kenjis) [reply system/Config/BaseConfig.php]: It has only `@var`, so it is okay.

https://github.com/codeigniter4/CodeIgniter4/blob/6f81e6e42e996eb1d0eacb2325600f74fefc9307/system/Config/BaseConfig.php#L56-L60

**Review by kenjis — APPROVED** [review]

File: tests/system/Config/BaseConfigTest.php [hunk 4]
```diff
@@ -265,32 +271,28 @@ public function testBadRegistrar(): void
...
-        $config              = new RegistrarConfig();
-        $config::$registrars = [];
+        $config = new RegistrarConfig();
         $this->setPrivateProperty($config, 'didDiscovery', false);
-
-        $method = $this->getPrivateMethodInvoker($config, 'registerProperties');
-        $method();
+        ($this->getPrivateMethodInvoker($config, 'registerProperties'))();
 
         $this->assertTrue($this->getPrivateProperty($config, 'didDiscovery'));
```
Comment (kenjis) [inline tests/system/Config/BaseConfigTest.php:297]: The following is easier to understand.

```diff
--- a/tests/system/Config/BaseConfigTest.php
+++ b/tests/system/Config/BaseConfigTest.php
@@ -290,9 +290,9 @@ final class BaseConfigTest extends CIUnitTestCase
         $locator->method('search')->with('Config/Registrar.php')->willReturn([]);
         Services::injectMock('locator', $locator);
 
+        $this->setPrivateProperty(RegistrarConfig::class, 'didDiscovery', false);
+
         $config = new RegistrarConfig();
-        $this->setPrivateProperty($config, 'didDiscovery', false);
-        ($this->getPrivateMethodInvoker($config, 'registerProperties'))();
 
         $this->assertTrue($this->getPrivateProperty($config, 'didDiscovery'));
     }
```

**Review by kenjis — APPROVED** [review]

**Review by MGatner — APPROVED** [review]

**Pull request merged**